app.R

```diff
@@ -70,49 +70,49 @@ render_subexp_ui <- function(subexp_name) {
 
 render_experiment_ui <- function(exp_name, full_cache_data) {
   
-  subexp_tabs <- lapply(names(full_cache_data[[ exp_name ]]), render_subexp_ui)
+  subexp_tabs <- lapply(names(full_cache_data[[ exp_name ]]), render_subexp_ui, exp_name)
   
   do.call(tabPanel, list(exp_name, div(class = 'push-down'),
     do.call(tabsetPanel, subexp_tabs))
   )
 }
 
 # render selection menu from available variables in a given subexperiment
-render_variable_menu <- function(subexp_name, output, full_cache_data) {
+render_variable_menu <- function(subexp_name, exp_name, output, full_cache_data) {
   
   variable_names <- names(full_cache_data[[ subexp_name ]][[ 'trait_data' ]])
   
-  output[[ paste0('variable_select_', subexp_name) ]] <- renderUI({
-    selectInput(paste0('selected_variable_', subexp_name), 'Variable', variable_names)
+  output[[ paste0('variable_select_', exp_name, '_', subexp_name) ]] <- renderUI({
+    selectInput(paste0('selected_variable_', exp_name, '_', subexp_name), 'Variable', variable_names)
   })
 }
 
 # render selection menu from available cultivars in a given subexperiment, for the selected variable
-render_cultivar_menu <- function(subexp_name, input, output, full_cache_data) {
+render_cultivar_menu <- function(subexp_name, exp_name, input, output, full_cache_data) {
   
-  output[[ paste0('cultivar_select_', subexp_name) ]] <- renderUI({
+  output[[ paste0('cultivar_select_', exp_name, '_', subexp_name) ]] <- renderUI({
     
-    req(input[[ paste0('selected_variable_', subexp_name) ]])
+    req(input[[ paste0('selected_variable_', exp_name, '_', subexp_name) ]])
     
-    trait_records <- full_cache_data[[ subexp_name ]][[ 'trait_data' ]][[ input[[ paste0('selected_variable_', subexp_name) ]] ]][[ 'traits' ]]
+    trait_records <- full_cache_data[[ subexp_name ]][[ 'trait_data' ]][[ input[[ paste0('selected_variable_', exp_name, '_', subexp_name) ]] ]][[ 'traits' ]]
     unique_cultivars <- unique(trait_records[[ 'cultivar_name' ]])
     
-    selectInput(paste0('selected_cultivar_', subexp_name), 'Cultivar', c('None', unique_cultivars))
+    selectInput(paste0('selected_cultivar_', exp_name, '_', subexp_name), 'Cultivar', c('None', unique_cultivars))
   })
 }
 
 # render box plot time series from trait records in a given subexperiment, for the selected variable
 # if a cultivar is selected, render line plot from trait records for that cultivar
-render_trait_plot <- function(subexp_name, input, output, full_cache_data) {
+render_trait_plot <- function(subexp_name, exp_name, input, output, full_cache_data) {
```

Comment:
A minor point, doesn’t affect the functionality but a simple place to think about making code more [DRY](url
https://en.m.wikipedia.org/wiki/Don%27t_repeat_yourself) ... the pattern `exp_name, '_', subexp_name` appears a dozen or more times ... is there a way to simplify this and just paste it together once?

Reply:
Yes, I think there is a better way to simplify this pattern. 

I have created a variable named `id_str` and set its value to `paste0(exp_name, '_', subexp_name)`  in `render_subexp_output` and `render_subexp_ui`.

Reply:
I added `id_str` as a parameter in some functions that require this string for input/output ids.